src/NLog/Internal/PropertySetter.cs

```diff
@@ -0,0 +1,119 @@
+//   this list of conditions and the following disclaimer. 
+// 
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution. 
+// 
+// * Neither the name of Jaroslaw Kowalski nor the names of its 
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission. 
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+namespace NLog.Internal
+{
+    using System;
+    using System.Reflection;
+    using NLog.Common;
+
+    internal interface IPropertySetter
+    {
+        bool SetPropertyValue(object instance, object value);
+    }
+
+    internal class PropertySetter : IPropertySetter
+    {
+        private readonly Type _objectType;
+        private readonly PropertyInfo _propertyInfo;
+        private ReflectionHelpers.LateBoundMethod _propertySetter;
+
+        public static IPropertySetter GeneratePropertySetter(Type objectType, string propertyName)
+        {
+            if (TryGetProperty(objectType, propertyName, out var propertyInfo))
+            {
+                return new PropertySetter(objectType, propertyInfo);
+            }
+
+            return null;
+        }
+
+        private PropertySetter(Type objectType, PropertyInfo propertyInfo)
+        {
+            _objectType = objectType;
+            _propertyInfo = propertyInfo;
+        }
+
+        public bool SetPropertyValue(object instance, object value)
+        {
+            if (instance == null)
+            {
+                throw new ArgumentNullException(nameof(instance));
+            }
+
+            if (_propertySetter != null)
+            {
+                _propertySetter.Invoke(instance, new[] { value });
+            }
+            else
+            {
+                // Generate compiled method if setter works without throwing exception
+                _propertyInfo.GetSetMethod().Invoke(instance, new[] { value });
+                _propertySetter = ReflectionHelpers.CreateLateBoundMethod(_propertyInfo.GetSetMethod());
```

Comment:
Maybe doing once `_propertyInfo.GetSetMethod()`? (sounds like an expensive method).

Maybe describe this class as "lazy compiled" (In class XML docs)

Reply:
Not expensive, but I can move it into a local variable for you.

Reply:
I like locals :)